engine/src/damage/damageable_layer.cpp

```diff
@@ -0,0 +1,250 @@
+        facet.AdjustPower(adjusted_percent);
+    }
+}
+
+void DamageableLayer::DealDamage( const CoreVector &attack_vector, Damage &damage, InflictedDamage& inflicted_damage ) {
+    int impacted_facet_index = GetFacetIndex(attack_vector);
+    facets[impacted_facet_index].DealDamage(damage, inflicted_damage);
+}
+
+
+void DamageableLayer::Destroy() {
+    for(Health& facet : facets) {
+        facet.health = 0;
+        facet.destroyed = true;
+    }
+}
+
+
+void DamageableLayer::Disable() {
+    for(Health& facet : facets) {
+        facet.health = 0;
+        facet.enabled = false;
+    }
+}
+
+// Used for nicer graphics when entering SPEC
+void DamageableLayer::GradualDisable(float percent) {
+    const float current_percent = facets[0].adjusted_health / facets[0].max_health;
+    AdjustPower(current_percent - percent);
+}
+
+
+void DamageableLayer::Enable() {
+    for(Health& facet : facets) {
+        facet.enabled = true;
+    }
+}
+
+// TODO: test
+// Boost shields to 150%
+void DamageableLayer::Enhance() {
+    for(Health& facet: facets) {
+        // Don't enhance armor and hull
+        if(!facet.regenerative) {
+            continue;
+        }
+
+        facet.health = facet.max_health * 1.5;
+    }
+}
+
+int DamageableLayer::GetFacetIndex(const CoreVector& attack_vector) {
+    // Convenience Variables
+    float i = attack_vector.i;
+    float j = attack_vector.j;
+    float k = attack_vector.k;
+
+    if(configuration == FacetConfiguration::one) {
+        return 0;
+    } else if(configuration == FacetConfiguration::two) {
+        if(k >= 0) {
+            return 0;
+        } else {
+            return 1;
+        }
+    } else if(configuration == FacetConfiguration::four) {
+
+    } else if(configuration == FacetConfiguration::six) {
+
+    } else if(configuration == FacetConfiguration::eight) {
+        if(i >= 0 && j >= 0 && k >= 0) { return 0; }
+        if(i < 0 && j >= 0 && k >= 0) { return 1; }
+        if(i >= 0 && j < 0 && k >= 0) { return 2; }
+        if(i < 0 && j < 0 && k >= 0) { return 3; }
+        if(i >= 0 && j >= 0 && k < 0) { return 4; }
+        if(i < 0 && j >= 0 && k < 0) { return 5; }
+        if(i >= 0 && j < 0 && k < 0) { return 6; }
+        if(i < 0 && j < 0 && k < 0) { return 7; }
+    }
+
+    // The default
+    return 0;
+}
+
+int DamageableLayer::GetFacetIndexByName(FacetName facet_name) {
+    // We ignore the parameter and just return the single facet
+    if(configuration == FacetConfiguration::one) {
+        return 0;
+    } else if(configuration == FacetConfiguration::two) {
+        return Find(facet_name, two_configuration, 2);
+    } else if(configuration == FacetConfiguration::four) {
+        return Find(facet_name, four_configuration, 4);
+    } else if(configuration == FacetConfiguration::six) {
+        return Find(facet_name, six_configuration, 6);
+    } else if(configuration == FacetConfiguration::eight) {
+        return Find(facet_name, eight_configuration, 8);
+    }
```

Comment:
A `switch` block would probably be better:
```C++
switch (configuration) {
    case FaceConfiguration::one:
        return 0;
        break;
    case FaceConfiguration::two:
        return Find(face_name, two_configuration, 2);
        break;
    ...
}
```
Alternatively using a map for the second two parameters would be just as good, possibly even better.
```C++
// map version would be something like:
if (configuration == FacetConfiguration::one) {
    return 0;
} else {
   // needs a check to make sure that the configuration is in the map
   // parameters is a pair with the configuration and integer value
   parameters = parameterMap[configuration];
   return Find(face_name, parameters.first, parameters.second);
}
```
Though there may be a better way to map the values out too.
Both would yield higher performance than the multiple IF-blocks.

Reply:
Sorry. For some reason I assumed enum class does not support switches. Will fix.

Reply:
Done.